File: apps/web/lib/read-selector.ts
// Mirror of the Go `readselector` package
// (apps/backend/internal/common/readselector). OMP's read tool embeds a
// line/range/mode selector in the path ("foo.go:43-94") and can reference
// several comma-joined files in a single read ("a.yaml:1-80,b.yaml:1-80"). The
// backend keeps a multi-file read's path raw; this splits it into one openable
// entry per file (selector stripped, line range parsed) so the chat can render a
// separate link per file. Keep in sync with the Go implementation and its tests.

export type ReadFileRef = {
  /** Bare, openable file path with any read selector stripped. */
  path: string;
  /** 1-based first line referenced by the selector, or 0 when none. */
  startLine: number;
  /** Contiguous line span for a closed range ("N-M" / "N+K"), else 0. */
  lineCount: number;
};

type LineSpec = { startLine: number; lineCount: number };

// atoi mirrors Go strconv.Atoi for our needs: an optionally-signed decimal
// integer only (no hex, exponent, or surrounding whitespace).
function atoi(s: string): number | null {
  if (!/^-?\d+$/.test(s)) return null;
  return Number(s);
}

// parseLineSpec parses a single line spec: "N", "N-", "N-M", or "N+K".
function parseLineSpec(seg: string): LineSpec | null {
  const plus = seg.indexOf("+");
  if (plus >= 0) {
    const start = atoi(seg.slice(0, plus));
    const count = atoi(seg.slice(plus + 1));
    if (start === null || count === null || start <= 0 || count < 0) return null;
    return { startLine: start, lineCount: count };
  }
  const dash = seg.indexOf("-");
  if (dash >= 0) {
    const start = atoi(seg.slice(0, dash));
    if (start === null || start <= 0) return null;
    const rest = seg.slice(dash + 1);
    if (rest === "") return { startLine: start, lineCount: 0 }; // "N-" open-ended
    const end = atoi(rest);
    if (end === null || end < start) return null;
    return { startLine: start, lineCount: end - start + 1 };
  }
  const start = atoi(seg);
  if (start === null || start <= 0) return null;
  return { startLine: start, lineCount: 0 };
}

// parseLineSpecList parses a comma-separated list ("5-16,960-973"); the first
// spec drives the result.
function parseLineSpecList(part: string): LineSpec | null {
  const segs = part.split(",");
  let first: LineSpec | null = null;
  for (let i = 0; i < segs.length; i++) {
    const spec = parseLineSpec(segs[i] ?? "");
    if (!spec) return null;
    if (i === 0) first = spec;
  }
  return first;
}

// parseReadSelector validates the selector tail (everything after the file's
// first ':'). Parts joined by ':' accept combos such as "2-4:raw"; the first
// line-spec encountered drives the range. Returns a zero range (still valid) for
// mode-only selectors like "raw".
function parseReadSelector(suffix: string): LineSpec | null {
  let result: LineSpec = { startLine: 0, lineCount: 0 };
  let gotLine = false;
  for (const part of suffix.split(":")) {
    if (part === "raw" || part === "conflicts") continue;
    const spec = parseLineSpecList(part);
    if (!spec) return null;
    if (!gotLine) {
      result = spec;
      gotLine = true;
    }
  }
  return result;
}

// splitOne mirrors Go readselector.Split for a single file reference. A colon in
// the file's final segment that isn't a Windows drive letter ("C:\...") begins
// the selector.
function splitOne(raw: string): ReadFileRef {
  const unchanged: ReadFileRef = { path: raw, startLine: 0, lineCount: 0 };
  let lastSep = raw.lastIndexOf("/");
  const back = raw.lastIndexOf("\\");
  if (back > lastSep) lastSep = back;
  const rel = raw.slice(lastSep + 1).indexOf(":");
  if (rel < 0) return unchanged;
  const colon = lastSep + 1 + rel;
  const isWindowsDrivePrefix = colon === 1 && raw.length >= 2 && /^[A-Za-z]$/.test(raw[0] ?? "");
  if (lastSep < 0 && isWindowsDrivePrefix) return unchanged;
  const base = raw.slice(0, colon);
  const suffix = raw.slice(colon + 1);
  if (base === "" || suffix === "") return unchanged;
  const parsed = parseReadSelector(suffix);
  if (!parsed) return unchanged;
  return { path: base, startLine: parsed.startLine, lineCount: parsed.lineCount };
}

// isStrongFile reports whether a comma segment is a self-contained file
// reference: it carries a read selector (splitOne strips it) or its basename has
// an extension. A segment that looks fileish only via a path separator ("src/foo"
// in the single filename "src/foo,bar.go") is NOT a file boundary, so a comma
// inside a filename is not mistaken for a multi-file bundle.
function isStrongFile(part: string): boolean {
  if (splitOne(part).path !== part) return true;
  const sep = Math.max(part.lastIndexOf("/"), part.lastIndexOf("\\"));
  const base = sep >= 0 ? part.slice(sep + 1) : part;
  return base.includes(".");
}

/**
 * Split an omp read path that may reference several comma-joined files into one
 * {@link ReadFileRef} per file. A single file (with or without a multi-range
 * selector) returns exactly one entry identical to the Go `Split`; multiple
 * files are only reported when every segment is a self-contained file reference
 * (it carries a read selector or its basename has an extension), so a comma
 * inside a directory name ("a,b/foo.go") or a filename ("src/foo,bar.go:1-20")
 * stays a single entry rather than splitting into phantom files.
 */
export function splitReadFiles(raw: string): ReadFileRef[] {
  const parts = raw.split(",");
  const files: ReadFileRef[] = [];
  let multi = true;
  parts.forEach((part, i) => {
    // A bare line-spec list or mode keyword is an extra range of the preceding
    // file, not a new file.
    const isContinuationRange =
      part === "raw" || part === "conflicts" || parseLineSpecList(part) !== null;
    if (i > 0 && isContinuationRange) return;
    if (!isStrongFile(part)) multi = false;
    files.push(splitOne(part));
  });
  if (multi && files.length > 1) return files;
  return [splitOne(raw)];
}

File: apps/backend/internal/common/readselector/readselector.go
// Package readselector parses Oh-My-Pi (omp) style read paths of the form
// "path:<selector>" into the bare file path plus the parsed line range.
//
// OMP's read tool embeds a line/range/mode selector in the path argument
// (e.g. "foo.go:43-94", "foo.go:50+150", "foo.go:5-16,960-973", "foo.go:raw",
// "foo.go:2-4:raw"). Kandev opens files by stat'ing the path, so a selector
// left on the path makes the open fail ("Failed to open file … no such file"
// for workspace files, or "path traversal detected" for absolute external
// files whose stat'd "<path>:<range>" simply doesn't exist).
//
// Both the ACP normalizer (so new file links are stored clean and the line
// range is surfaced separately) and the file-open boundary (so already-persisted
// links and any normalization gap still open) call Split — a single source of
// truth shared by both.
package readselector

import (
	"strconv"
	"strings"
)

// Split splits "path:<selector>" into the bare file path plus the parsed line
// range. startLine is the first line referenced by the selector; lineCount is
// the contiguous span when a closed range is given ("N-M" or "N+K"), and 0 when
// the selector is open-ended ("N", "N-") or carries no line numbers ("raw").
// When no valid selector is present the path is returned unchanged with zero
// line info, so paths from non-omp agents (which never match the grammar) and
// ordinary filenames are unaffected.
func Split(raw string) (path string, startLine, lineCount int) {
	// Only inspect colons in the final path segment so a directory component
	// that legitimately contains a colon is never mistaken for a selector.
	// Consider both '/' and '\\' so Windows paths — and their drive-letter
	// colon, e.g. "C:\\dir\\file.go" — are not misread as a selector boundary.
	lastSep := strings.LastIndexByte(raw, '/')
	if b := strings.LastIndexByte(raw, '\\'); b > lastSep {
		lastSep = b
	}
	rel := strings.IndexByte(raw[lastSep+1:], ':')
	if rel < 0 {
		return raw, 0, 0
	}
	colon := lastSep + 1 + rel
	if lastSep < 0 && isWindowsDrivePrefix(raw, colon) {
		return raw, 0, 0
	}
	base, suffix := raw[:colon], raw[colon+1:]
	if base == "" || suffix == "" {
		return raw, 0, 0
	}
	start, count, ok := parseReadSelector(suffix)
	if !ok {
		return raw, 0, 0
	}
	return base, start, count
}

// File is one file reference parsed out of a (possibly multi-file) read path.
type File struct {
	Path      string
	StartLine int
	LineCount int
}

// SplitFiles splits an omp read path that may reference several comma-joined
// files ("a.yaml:1-80,b.yaml:1-80") into one File per file, each with its own
// parsed line range. A comma segment that is purely a line-spec list ("960-973")
// or a mode keyword ("raw"/"conflicts") is treated as an additional range of the
// preceding file — so a single file's multi-range selector ("main.go:5-16,40-80")
// stays one File, matching Split (first range wins).
//
// Multiple files are only reported when every segment is a self-contained file
// reference — it carries a read selector or its basename has an extension.
// Otherwise it returns the single Split result, so single-file behavior is never
// disturbed: a comma inside a directory name ("a,b/foo.go") or inside a filename
// ("src/foo,bar.go:1-20") stays one File rather than splitting into phantoms.
func SplitFiles(raw string) []File {
	parts := strings.Split(raw, ",")
	files := make([]File, 0, len(parts))
	multi := true
	for i, part := range parts {
		if i > 0 && isContinuationRange(part) {
			continue
		}
		if !isStrongFile(part) {
			multi = false
		}
		path, start, count := Split(part)
		files = append(files, File{Path: path, StartLine: start, LineCount: count})
	}
	if multi && len(files) > 1 {
		return files
	}
	path, start, count := Split(raw)
	return []File{{Path: path, StartLine: start, LineCount: count}}
}

// isContinuationRange reports whether a comma segment is an extra range of the
// preceding file rather than a new file: a bare line-spec list or a mode keyword.
func isContinuationRange(part string) bool {
	if part == "raw" || part == "conflicts" {
		return true
	}
	_, _, ok := parseLineSpecList(part)
	return ok
}

// isStrongFile reports whether a comma segment is a self-contained file
// reference: it carries a read selector (Split strips it) or its basename has an
// extension. A segment that looks fileish only via a path separator ("src/foo"
// in the single filename "src/foo,bar.go") is NOT a file boundary, so a comma
// inside a filename is not mistaken for a multi-file bundle.
func isStrongFile(part string) bool {
	if path, _, _ := Split(part); path != part {
		return true
	}
	base := part
	if i := strings.LastIndexAny(part, `/\`); i >= 0 {
		base = part[i+1:]
	}
	return strings.Contains(base, ".")
}

func isWindowsDrivePrefix(raw string, colon int) bool {
	return colon == 1 && len(raw) >= 2 && isASCIIAlpha(raw[0])
}

func isASCIIAlpha(b byte) bool {
	return (b >= 'A' && b <= 'Z') || (b >= 'a' && b <= 'z')
}

// parseReadSelector validates the full selector tail (everything after the
// first ':' in the final segment). Selector parts are joined by ':' so combos
// such as "2-4:raw" and "raw:2-4" are accepted; each part must be a line-spec
// list or a recognized mode keyword. The first line-spec encountered drives the
// returned startLine/lineCount.
func parseReadSelector(suffix string) (startLine, lineCount int, ok bool) {
	gotLine := false
	for _, part := range strings.Split(suffix, ":") {
		if part == "raw" || part == "conflicts" {
			continue
		}
		s, c, valid := parseLineSpecList(part)
		if !valid {
			return 0, 0, false
		}
		if !gotLine {
			startLine, lineCount, gotLine = s, c, true
		}
	}
	return startLine, lineCount, true
}

// parseLineSpecList parses a comma-separated list of line specs ("5-16,960-973")
// and reports the start/count of the first spec.
func parseLineSpecList(part string) (startLine, lineCount int, ok bool) {
	segs := strings.Split(part, ",")
	for i, seg := range segs {
		s, c, valid := parseLineSpec(seg)
		if !valid {
			return 0, 0, false
		}
		if i == 0 {
			startLine, lineCount = s, c
		}
	}
	return startLine, lineCount, true
}

// parseLineSpec parses a single line spec: "N", "N-", "N-M", or "N+K".
func parseLineSpec(seg string) (startLine, lineCount int, ok bool) {
	if i := strings.IndexByte(seg, '+'); i >= 0 {
		start, errA := strconv.Atoi(seg[:i])
		count, errB := strconv.Atoi(seg[i+1:])
		if errA != nil || errB != nil || start <= 0 || count < 0 {
			return 0, 0, false
		}
		return start, count, true
	}
	if i := strings.IndexByte(seg, '-'); i >= 0 {
		start, errA := strconv.Atoi(seg[:i])
		if errA != nil || start <= 0 {
			return 0, 0, false
		}
		rest := seg[i+1:]
		if rest == "" { // "N-" — open-ended
			return start, 0, true
		}
		end, errB := strconv.Atoi(rest)
		if errB != nil || end < start {
			return 0, 0, false
		}
		return start, end - start + 1, true
	}
	start, err := strconv.Atoi(seg)
	if err != nil || start <= 0 {
		return 0, 0, false
	}
	return start, 0, true
}
